File: planning/collisionChecker.m
function collisionReport = collisionChecker(currentPose, currentVelocity, currentPath, detectedObjects, cfg)
% COLLISIONCHECKER Assesses collision risk and computes Time-To-Collision (TTC) along vehicle path.
%
% Risk Levels:
%   "SAFE"     - No obstacles threatening vehicle path
%   "WARNING"  - Obstacle approaching corridor or within warning threshold
%   "CRITICAL" - Imminent collision risk (TTC < threshold or obstacle blocking path)
%
% Inputs:
%   currentPose     - [x y theta] Current ego pose (world frame)
%   currentVelocity - Current ego speed (m/s)
%   currentPath     - Active PlannedPath struct (containing .waypoints)
%   detectedObjects - Array of detected object structs
%   cfg             - System configuration struct
%
% Output:
%   collisionReport - Struct containing:
%                     .riskLevel        - "SAFE" | "WARNING" | "CRITICAL"
%                     .minDistance      - Minimum distance to obstacle (m)
%                     .minTTC           - Minimum Time-To-Collision (s) (inf if none)
%                     .relVelocity      - Relative closing velocity (m/s)
%                     .criticalObstacle - Struct of threatening obstacle
%                     .needsReplan      - Logical flag indicating replanning required
%                     .reason           - Human-readable explanation string

    if nargin < 5 || isempty(cfg)
        cfg = config();
    end

    ttcCrit = cfg.risk.ttcCritical;
    ttcWarn = cfg.risk.ttcWarning;
    distCrit = cfg.risk.distCritical;
    distWarn = cfg.risk.distWarning;
    corridorWidth = cfg.risk.lateralClearance;

    numObs = numel(detectedObjects);

    % Default SAFE report
    collisionReport = struct(...
        'riskLevel', "SAFE", ...
        'minDistance', inf, ...
        'minTTC', inf, ...
        'relVelocity', 0.0, ...
        'criticalObstacle', struct('id', 0, 'class', "none"), ...
        'needsReplan', false, ...
        'reason', "Path is clear and safe.");

    if numObs == 0 || isempty(currentPath) || ~isfield(currentPath, 'waypoints') || isempty(currentPath.waypoints)
        return;
    end

    egoPos = currentPose(1:2);
    egoHeading = currentPose(3);
    egoVelVec = [currentVelocity * cos(egoHeading), currentVelocity * sin(egoHeading)];

    minDist = inf;
    minTTC = inf;
    closingSpeed = 0.0;
    highestRisk = "SAFE";
    criticalObj = struct('id', 0, 'class', "none");
    needsReplan = false;
    reasonStr = "Path clear";

    waypoints = currentPath.waypoints;
    numWp = size(waypoints, 1);

    % Check each detected obstacle
    for i = 1:numObs
        obj = detectedObjects(i);
        obsPos = obj.worldPosition;
        obsVel = obj.worldVelocity;
        obsRadius = obj.radius;
        dist = norm(obsPos - egoPos);

        if dist < minDist
            minDist = dist;
        end

        % Relative position and closing velocity
        relPos = obsPos - egoPos;
        relVel = egoVelVec - obsVel; % Positive means ego is closing in on obstacle
        currClosingSpeed = dot(relPos, relVel) / (dist + 1e-5);

        % Compute Time-To-Collision (TTC)
        ttc = inf;
        if currClosingSpeed > 0.1
            ttc = (dist - (obsRadius + cfg.vehicle.length / 2)) / currClosingSpeed;
            if ttc < 0
                ttc = 0.0;
            end
        end

        if ttc < minTTC
            minTTC = ttc;
            closingSpeed = currClosingSpeed;
        end

        % Check if obstacle intersects the active planned path corridor ahead
        % Lookahead along waypoints within next 15 meters
        pathIntersects = false;
        for w = 1:min(numWp, 40)
            wp = waypoints(w, :);
            distWpToEgo = norm(wp - egoPos);
            if distWpToEgo > 25.0
                break;
            end

            distToPath = norm(wp - obsPos);
            if distToPath <= (obsRadius + corridorWidth)
                pathIntersects = true;
                break;
            end
        end

        % Risk Evaluation Logic
        objRisk = "SAFE";
        if dist <= distCrit || ttc <= ttcCrit || (pathIntersects && dist <= (distCrit * 2.0))
            objRisk = "CRITICAL";
            needsReplan = true;
        elseif dist <= distWarn || ttc <= ttcWarn || pathIntersects
            objRisk = "WARNING";
            if pathIntersects
                needsReplan = true;
            end
        end

        % Update highest risk found
        if (objRisk == "CRITICAL")
            highestRisk = "CRITICAL";
            criticalObj = obj;
            closingSpeed = currClosingSpeed;
            reasonStr = sprintf("%s detected %.1fm ahead (TTC: %.1fs) - Collision imminent!", ...
                upper(obj.class), dist, min(99.9, ttc));
            break; % Critical is maximum priority
        elseif (objRisk == "WARNING") && (highestRisk ~= "CRITICAL")
            highestRisk = "WARNING";
            criticalObj = obj;
            closingSpeed = currClosingSpeed;
            reasonStr = sprintf("%s in proximity (%.1fm, TTC: %.1fs) - Cautious navigation", ...
                upper(obj.class), dist, min(99.9, ttc));
        end
    end

    collisionReport.riskLevel = highestRisk;
    collisionReport.minDistance = minDist;
    collisionReport.minTTC = minTTC;
    collisionReport.relVelocity = closingSpeed;
    collisionReport.criticalObstacle = criticalObj;
    collisionReport.needsReplan = needsReplan;
    collisionReport.reason = reasonStr;
end

File: smart-autonomous-vehicle/backend/planning/collision_checker.py
"""
Collision Risk Engine & Time-To-Collision (TTC) Evaluator.
"""

from dataclasses import dataclass
from typing import Tuple, List, Optional, Dict, Any
import numpy as np

try:
    from ..config import Config, get_config
    from ..perception.object_detector import DetectedObject
    from .adaptive_planner import PlannedPath
    from .global_path import GlobalPath
except (ImportError, ValueError):
    try:
        from backend.config import Config, get_config
        from backend.perception.object_detector import DetectedObject
        from backend.planning.adaptive_planner import PlannedPath
        from backend.planning.global_path import GlobalPath
    except (ImportError, ValueError):
        from config import Config, get_config
        from perception.object_detector import DetectedObject
        from planning.adaptive_planner import PlannedPath
        from planning.global_path import GlobalPath


@dataclass
class CollisionReport:
    risk_level: str                 # "SAFE" | "WARNING" | "CRITICAL"
    min_distance: float             # Proximity to nearest hazard (m)
    min_ttc: float                  # Minimum Time-To-Collision (s) (inf if none)
    rel_velocity: float             # Relative closing velocity (m/s)
    critical_obstacle: Optional[DetectedObject]
    needs_replan: bool              # Flag to trigger replanning
    reason: str                     # Human-readable explanation string

    def to_dict(self) -> Dict[str, Any]:
        return {
            "risk_level": self.risk_level,
            "min_distance": round(self.min_distance, 2) if not np.isinf(self.min_distance) else 999.0,
            "min_ttc": round(self.min_ttc, 2) if not np.isinf(self.min_ttc) else 999.0,
            "rel_velocity": round(self.rel_velocity, 2),
            "critical_obstacle": self.critical_obstacle.to_dict() if self.critical_obstacle else None,
            "needs_replan": self.needs_replan,
            "reason": self.reason,
        }


def collision_checker(
    current_pose: Tuple[float, float, float],
    current_velocity: float,
    current_path: Any,
    detected_objects: List[DetectedObject],
    cfg: Optional[Config] = None,
) -> CollisionReport:
    """
    Assesses collision risk, calculates TTC, and identifies whether path replanning is needed.
    """
    if cfg is None:
        cfg = get_config()

    ttc_crit = cfg.risk.ttc_critical
    ttc_warn = cfg.risk.ttc_warning
    dist_crit = cfg.risk.dist_critical
    dist_warn = cfg.risk.dist_warning
    corridor_w = cfg.risk.lateral_clearance

    if not detected_objects or current_path is None or len(getattr(current_path, "waypoints", [])) == 0:
        return CollisionReport(
            risk_level="SAFE",
            min_distance=float("inf"),
            min_ttc=float("inf"),
            rel_velocity=0.0,
            critical_obstacle=None,
            needs_replan=False,
            reason="Path is clear and safe.",
        )

    ego_pos = np.array(current_pose[:2], dtype=float)
    ego_theta = current_pose[2]
    ego_vel_vec = np.array([current_velocity * np.cos(ego_theta), current_velocity * np.sin(ego_theta)])

    waypoints = current_path.waypoints
    min_dist = float("inf")
    min_ttc = float("inf")
    closing_speed = 0.0
    highest_risk = "SAFE"
    critical_obj: Optional[DetectedObject] = None
    needs_replan = False
    reason_str = "Path clear."

    for obj in detected_objects:
        obs_pos = np.array(obj.world_position, dtype=float)
        obs_vel = np.array(obj.world_velocity, dtype=float)
        obs_radius = obj.radius

        dist = float(np.linalg.norm(obs_pos - ego_pos))
        if dist < min_dist:
            min_dist = dist

        rel_pos = obs_pos - ego_pos
        rel_vel = ego_vel_vec - obs_vel
        curr_closing_speed = float(np.dot(rel_pos, rel_vel) / (dist + 1e-5))

        ttc = float("inf")
        if curr_closing_speed > 0.1:
            ttc = max(0.0, (dist - (obs_radius + cfg.vehicle.length / 2.0)) / curr_closing_speed)

        if ttc < min_ttc:
            min_ttc = ttc
            closing_speed = curr_closing_speed

        # Check if obstacle intersects planned corridor ahead
        path_intersects = False
        for wp in waypoints[:40]:
            dist_wp_to_ego = float(np.linalg.norm(wp - ego_pos))
            if dist_wp_to_ego > 25.0:
                break
            if float(np.linalg.norm(wp - obs_pos)) <= (obs_radius + corridor_w):
                path_intersects = True
                break

        # Risk state assignment
        obj_risk = "SAFE"
        if (ttc <= ttc_crit) or (path_intersects and dist <= (dist_crit * 1.5)):
            obj_risk = "CRITICAL"
            needs_replan = True
        elif dist <= dist_crit and path_intersects:
            obj_risk = "CRITICAL"
            needs_replan = True
        elif dist <= dist_warn or ttc <= ttc_warn or path_intersects:
            obj_risk = "WARNING"
            if path_intersects:
                needs_replan = True

        # If vehicle has stopped and detour path is clear around obstacle, allow crawl bypass
        if current_velocity < 0.3 and not path_intersects and obj_risk == "CRITICAL":
            obj_risk = "WARNING"

        if obj_risk == "CRITICAL":
            highest_risk = "CRITICAL"
            critical_obj = obj
            closing_speed = curr_closing_speed
            reason_str = f"{obj.object_class.upper()} detected {dist:.1f}m ahead (TTC: {min(99.9, ttc):.1f}s) - Collision imminent!"
            break
        elif obj_risk == "WARNING" and highest_risk != "CRITICAL":
            highest_risk = "WARNING"
            critical_obj = obj
            closing_speed = curr_closing_speed
            reason_str = f"{obj.object_class.upper()} in proximity ({dist:.1f}m, TTC: {min(99.9, ttc):.1f}s) - Caution required."

    return CollisionReport(
        risk_level=highest_risk,
        min_distance=min_dist,
        min_ttc=min_ttc,
        rel_velocity=closing_speed,
        critical_obstacle=critical_obj,
        needs_replan=needs_replan,
        reason=reason_str,
    )
